src/onnx/parse_if.cpp

```diff
@@ -65,13 +67,103 @@ struct parse_if : op_parser<parse_if>
         {
             MIGRAPHX_THROW("PARSE_IF: " + info.name +
-                           " then and else sub_grahps must have same output shapes!");
+                           " then and else sub_grahps must have same output type! " +
+                           then_out_shapes.at(0).type_string() + " vs " +
+                           else_out_shapes.at(0).type_string());
+        }
+
+        if(not then_out_shapes.at(0).dynamic() && not else_out_shapes.at(0).dynamic())
+        {
+            auto then_lens = then_out_shapes.at(0).lens();
+            auto else_lens = else_out_shapes.at(0).lens();
+
+            // Throw error if both branches have zero output shapes. Not possible for static inputs
+            if(then_lens.empty() && else_lens.empty())
+            {
+                throw_shapes();
+            }
+
+            auto handle_empty_branch = [](module_ref& mdl, const shape& out_shape) {
+                auto outline_ins = mdl->add_outline(out_shape);
+                mdl->replace_return({outline_ins});
+                return out_shape.lens();
+            };
+
+            // Handle one empty branch by setting output identical to the other
+            // need to update the then_shape before we do further checks
+            if(then_lens.empty())
+            {
+                then_lens = handle_empty_branch(then_mdl, else_out_shapes.at(0));
+            }
+            else if(else_lens.empty())
+            {
+                else_lens = handle_empty_branch(else_mdl, then_out_shapes.at(0));
+            }
+
+            // check equivilant length dims, and (x1,x2,.., xn, 1) == (x1,x2,..,xn)
+            int dim_delta = abs((static_cast<int>(then_lens.size() - else_lens.size())));
+
+            if(dim_delta == 1)
+            {
+                auto all_but_last_dims_equal = [](std::vector<size_t>& lens_a,
+                                                  std::vector<size_t>& lens_b) {
+                    if(lens_a.size() <= lens_b.size())
+                    {
+                        return equal(lens_a.begin(), lens_a.end(), lens_b.begin());
+                    }
+                    else
+                    {
+                        return equal(lens_b.begin(), lens_b.end(), lens_a.begin());
+                    }
+                };
+
+                // make sure dims are equivalent in static shapes
+                if(not all_but_last_dims_equal(then_lens, else_lens))
+                {
+                    throw_shapes();
+                }
+
+                auto unsqueeze_last_op = [](module_ref& mdl, const std::vector<size_t>& out_shape) {
+                    auto convert_ins = mdl->add_instruction(
+                        make_op("unsqueeze", {{"axes", {out_shape.size() - 1}}}),
+                        std::prev(std::prev(mdl->end())));
+                    mdl->replace_return({convert_ins});
+                    mdl->remove_instruction({std::prev(convert_ins)});
```

Comment:
DeadCode elimination should take care of removing this.

Reply:
is there a better way to do this? It was odd since I was somehow getting two return statements surprisingly when using replace_return

Reply:
Resolved from another comment.